splink/input_column.py

```diff
@@ -36,7 +68,51 @@ def __init__(self, name, tf_adjustments=False, settings_obj=None, sql_dialect=No
         else:
             self._sql_dialect = None
 
-        self._name_needs_escaping = _detect_if_name_needs_escaping(name)
+        if self._sql_dialect == "spark":
+            quote_character = "`"
+        else:
+            quote_character = '"'
+
+        # Quote SQL keywords
+        if name in ("group", "index"):
+            name = f"{quote_character}{name}{quote_character}"
+
+        self.input_name = name
+
+        self.input_name_as_tree = self.parse_input_name_to_sqlglot_tree()
+        self.quote()
+
+    def quote(self):
+        for identifier in self.input_name_as_tree.find_all(exp.Identifier):
+            identifier.set(arg="quoted", value=True)
+
+    def parse_input_name_to_sqlglot_tree(self):
+
+        # Cases that could occur for self.input_name:
+        # SUR name  -> parses to 'alias column identifier identifier'
+        # first and surname -> parses to 'and column column identifier identifier'
+        # a b c -> parse error
+        # "SUR name" -> parses to 'column identifier'
+        # geocode['lat'] -> parsees to bracket column literal identifier
+        # geocode[1] -> parsees to bracket column literal identifier
+
+        # Note we don't expect SUR name[1] since the user should have quoted this
+
+        try:
+            tree = sqlglot.parse_one(self.input_name, read=self._sql_dialect)
+        except ParseError:
+            tree = sqlglot.parse_one(f'"{self.input_name}"', read=self._sql_dialect)
+
+        tree_signature = sqlglot_tree_signature(tree)
+        valid_signatures = ["column identifier", "bracket column literal identifier"]
+
+        if tree_signature in valid_signatures:
+            return tree
+        else:
+            # e.g. SUR name parses to 'alias column identifier identifier'
+            # but we want "SUR name"
+            tree = sqlglot.parse_one(f'"{self.input_name}"', read=self._sql_dialect)
+            return tree
```

Comment:
This looks really good.

It also allows valid column names to sneak through in languages such as spark, that are in turn invalid in other dialects. 

For example:
```python
name, _sql_dialect = 'lat_long["lat"]', "spark" # << valid
parse_input_name_to_sqlglot_tree(name, _sql_dialect).sql() # << works

name, _sql_dialect = 'lat_long["lat"]', "duckdb" # << invalid
parse_input_name_to_sqlglot_tree(name, _sql_dialect).sql() # << breaks
```

Reply:
Hadn't thought of that!